### PR TITLE
feat(fused_moe): Support W4A8 MXFP4 MoE on Hopper (SM90)

### PR DESCRIPTION
**Pull request by yuan-luo**
## Description

Adds a CuTe DSL **W4A8 MXFP4** grouped GEMM for Hopper SM90 — FP8 e4m3 activation × MXFP4 weight (FP4 e2m1 + UE8M0 per-32-K-block scale) — with fused per-token gather, fused token scatter, and a fused SwiGLU epilogue for MoE. SM90 has no native FP4, so the weight is TMA-loaded as packed `Uint8 [N, K/2]` and dequantized to FP8 in-kernel before the FP8 wgmma.

- **Real MXFP4 beats plain FP16-dense by 1.30×** in the decode (weight-bandwidth-bound) regime, the W4A8 serving sweet spot.
- A full W4A8 MoE built from this kernel runs **~1.41× faster than cutlass W4A16** MoE end-to-end (avg over 2k–16k tokens; 1.47× at 16k).

## Modifications

- `flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_grouped_gemm_sm90.py` — persistent grouped GEMM kernel + the `w4a8_mxfp4_grouped_gemm` public API.
- `flashinfer/fused_moe/cute_dsl/moe_reduce_triton.py` — `moe_reduce` / `build_reduce_index`, a Triton non-atomic finalize for the top_k≥2 MoE path.
- Tests: `test_w4a8_mxfp4_moe_e2e.py` (16/16), `test_w4a8_swiglu.py` (3/3), `test_w4a8_moe_reduce.py` (12/12).
- Exports via `flashinfer.fused_moe` (cute-dsl/triton-conditional).

## Highlights

- **FP4→FP8 dequant, 5.6× over a scalar baseline**: vectorized swizzle-aware cotiled copy (2.16×) → `prmt` byte-LUT replacing per-nibble ALU (3.3×) → real UE8M0 scale folded into a per-group LUT, compile-time `%2` cached.
- **Shape-aware tile**: an ncu profile showed the GEMM is latency-bound at 7.8% occupancy (not MMA-bound). A tile sweep found the **largest** `(128,256)` cooperative tile (2 MMA warpgroups) fastest — more warps/CTA hide the dequant→smem→wgmma latency. The default auto-selects it when N % 256 == 0, else `(128,128)`.
- **Fused token scatter**: `top_k==1` uses a vectorized `v2.f32` plain store with per-pair index reuse; `top_k≥2` un-fuses to a plain store + `moe_reduce` (a fused atomicAdd is latency-bound — un-fusing to a memory-bound pass is **~2.6×** faster).
- **Fused SwiGLU epilogue** (`swiglu=True`): with gate/up interleaved in the weights, each pair lands in one thread's adjacent WGMMA registers, so GEMM1 writes `silu(gate)*up` to FP8 `[M, I]` directly — no `[M, 2I]` FP16 round-trip / requant kernel (1.32×→1.41×).
- **Host-overhead caching**: compiled-kernel + GPU-info + type-spec dummies + operand-metadata memoization bring repeat calls from **9180 ms → 0.18 ms**.
- **bf16/fp16/fp32 output** + crash-safe `@flashinfer_api` logging.

## Performance

**W4A8 MXFP4 vs plain FP16-dense** (H100):

| Regime | FP16-dense | W4A8 | Speedup |
|--------|-----------:|-----:|--------:|
| decode, real scale (32 experts × M=32, K=4096, N=2048) | 0.193 ms | **0.148 ms** | **1.30×** |
| compute, const scale (4×512×2048×4096) | ~500k GFLOPS | 643k | 1.29× |

**Full W4A8 MoE vs cutlass W4A16** (e=8, top_k=1, H=4096, I=1024, both incl. SwiGLU):

| tokens | W4A8 MoE | cutlass W4A16 | speedup |
|-------:|---------:|--------------:|--------:|
| 2048   | 196 µs   | 258 µs        | 1.31× |
| 8192   | 530 µs   | 776 µs        | 1.46× |
| 16384  | 977 µs   | 1439 µs       | 1.47× |
| **avg**| —        | —             | **1.41×** |

## Testing

- `test_w4a8_mxfp4_moe_e2e.py`: **16/16** — per-group, fused-scatter, and fused-gather paths × {FP16, BF16} × {64,128 tokens} × {top_k 1,2}, vs golden reference, on H100.
- `test_w4a8_swiglu.py`: **3/3** — fused SwiGLU epilogue vs torch `silu(gate)*up`.
- `test_w4a8_moe_reduce.py`: **12/12** — `moe_reduce` vs torch reduction over {top_k 1,2,3} × {fp16, bf16}.
- `ruff check` + `ruff format` clean.

## Notes

- Requires Hopper SM90 and the `nvidia-cutlass-dsl` package (CuTe DSL JIT). Tests skip on non-SM90. `moe_reduce` additionally requires `triton`.
- AI-assisted implementation.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * W4A8 MXFP4 Mixture-of-Experts (MoE) kernels optimized for NVIDIA Hopper (SM90), including fused SwiGLU, interleaving helpers, and multi‑top-k finalize support.
  * Triton-based MoE reduction and index-building routines for routed accumulation; optional clamped SwiGLU parameters per expert.
  * New fused MoE trace/template for the CuTe-DSL schema.

* **Tests**
  * Added gated CUDA/SM90 correctness and end-to-end tests covering fused paths, reductions, SwiGLU variants, and empty-batch handling.

### CODE REVIEW

File: flashinfer/fused_moe/cute_dsl/moe_reduce_triton.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        h = tl.program_id(1) * BLOCK_H + tl.arange(0, BLOCK_H)
+        mask = h < H
+        acc = tl.zeros([BLOCK_H], dtype=tl.float32)
+        for s in range(TOPK):
+            row = tl.load(idx_ptr + t * TOPK + s).to(tl.int64)
+            valid = row >= 0
+            w = tl.load(w_ptr + t * TOPK + s)
+            v = tl.load(c_ptr + row * H + h, mask=mask & valid, other=0.0).to(
+                tl.float32
+            )
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/moe_reduce_triton.py:71]: decline.
I don't think this is needed. `tl.load(..., mask=mask & valid, other=0.0)` is a predicated load — masked-off lanes never dereference the address, so computing a negative/out-of-range offset for them is safe and is the standard Triton idiom (same as any boundary-masked load like `tl.load(p + offs, mask=offs < N)`); predicated-off loads don't fault on CUDA. Also, in the current code path `permuted_idx` comes from `build_reduce_index` (argsort over a full routing), so indices are always ≥ 0 — the `-1` value is only the optional masked-expert case, which is already handled by `mask & valid` + `other=0.0`.

File: flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    and the fused epilogue bakes each as one compile-time const, so require uniformity
+    rather than silently using expert 0's value."""
+    if t is None:
+        return None
+    flat = t.reshape(-1)
+    if not bool(torch.all(flat == flat[0])):
+        raise NotImplementedError(
+            f"w4a8_mxfp4_moe requires a uniform per-expert {name} (got non-uniform values)"
+        )
+    return float(flat[0].item())
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py:146]: Good catch, adopted.

File: flashinfer/fused_moe/cute_dsl/moe_reduce_triton.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    device: Optional[torch.device] = None,
+) -> "tuple[torch.Tensor, torch.Tensor]":
+    """From per-group route maps (local row -> token) + weights, build the
+    ``(permuted_idx, topk_scales)`` of shape ``[num_tokens, top_k]`` that ``moe_reduce``
+    consumes. Assumes the per-group GEMM2 outputs are concatenated in group order (group
+    g's rows occupy ``[offset_g, offset_g + len(route_maps[g]))`` of ``permuted_out``).
+    Every token must appear exactly ``top_k`` times across the route maps.
+    """
+    device = device or route_maps[0].device
+    all_tok = torch.cat([r.long() for r in route_maps])  # [num_permuted]
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/moe_reduce_triton.py:238]: Good catch, the empty case is real. Adopted.

File: flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        raise ValueError(f"unsupported output_dtype {output_dtype}")
+    _check_unsupported(fc1_expert_biases, fc2_expert_biases)
+    # Clamped SwiGLU (GPT-OSS / DeepSeek-V4): a uniform per-expert (alpha, beta, limit)
+    # passed straight into the fused GEMM1 epilogue. None limit => plain silu(gate)*up.
+    alpha = _uniform_scalar(swiglu_alpha, "swiglu_alpha")
+    beta = _uniform_scalar(swiglu_beta, "swiglu_beta")
+    limit = _uniform_scalar(swiglu_limit, "swiglu_limit")
+
+    dev = input.device
+    T, H = input.shape
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py:258]: Already addressed in the latest revision, this is the same empty-batch case.

File: tests/moe/test_w4a8_moe_reduce.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    all_tok = torch.cat([r.long() for r in route])
+    all_w = torch.cat([w.float() for w in wts]).unsqueeze(1)
+    ref = torch.zeros(num_tokens, H, dtype=torch.float32, device=c_all.device)
+    ref.index_add_(0, all_tok, all_w * c_all.float())
+    return ref
+
+
+@pytest.mark.skipif(
+    not _OK or not torch.cuda.is_available(), reason="needs triton+CUDA"
+)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_moe_reduce.py:40]: `moe_reduce` is a standalone Triton kernel (a non-atomic gather-weighted-sum finalize) — it's not the SM90-only grouped GEMM and uses no Hopper-specific features (no wgmma/TMA/async), so it runs on any CUDA GPU with Triton.

File: flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    needs (row ``2j`` = gate_j, ``2j+1`` = up_j). Applies to both the packed weight
+    ``[E, 2I, H/2]`` and its scale ``[E, 2I, H/32]`` (same row permutation)."""
+    two_i = fc1_stacked.shape[1]
+    assert two_i % 2 == 0
+    i = two_i // 2
+    perm = torch.empty(two_i, dtype=torch.long, device=fc1_stacked.device)
+    perm[0::2] = torch.arange(i, device=fc1_stacked.device)  # gate -> even rows
+    perm[1::2] = torch.arange(i, 2 * i, device=fc1_stacked.device)  # up   -> odd rows
+    return fc1_stacked[:, perm].contiguous(), fc1_scale_stacked[:, perm].contiguous()
+
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py:118]: Added the decorator.

File: tests/moe/test_w4a8_mxfp4_moe_e2e.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+        f"|y|={y.abs().mean().item():.4f} |ref|={y_ref.abs().mean().item():.4f}"
+    )
+    return y, y_ref, frac_bad
+
+
+def _sm90():
+    if not torch.cuda.is_available():
+        return False
+    major, _ = torch.cuda.get_device_capability()
+    return major == 9
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_mxfp4_moe_e2e.py]: Fixed.

File: flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    flat = t.reshape(-1)
+    if not bool(torch.all(flat == flat[0])):
+        raise NotImplementedError(
+            f"w4a8_mxfp4_moe requires a uniform per-expert {name} (got non-uniform values)"
+        )
+    val = float(flat[0].item())
+    if len(_UNIFORM_SCALAR_CACHE) >= 256:
+        _UNIFORM_SCALAR_CACHE.clear()
+    _UNIFORM_SCALAR_CACHE[key] = val
+    return val
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py]: I've removed the cache rather than patch the key.

File: tests/moe/test_w4a8_moe_reduce.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+try:
+    from flashinfer.fused_moe.cute_dsl.moe_reduce_triton import (
+        build_reduce_index,
+        moe_reduce,
+    )
+
+    _OK = True
+except Exception:
+    _OK = False
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_moe_reduce.py:27]: Done.

File: tests/moe/test_w4a8_mxfp4_moe_e2e.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    """Import the callable from the kernel module (path-insensitive)."""
+    here = os.path.dirname(os.path.abspath(__file__))
+    repo = os.path.abspath(os.path.join(here, "..", ".."))
+    if repo not in sys.path:
+        sys.path.insert(0, repo)
+    from flashinfer.fused_moe.cute_dsl.w4a8_mxfp4_grouped_gemm_sm90 import (
+        w4a8_mxfp4_grouped_gemm,
+    )
+
+    return w4a8_mxfp4_grouped_gemm
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_mxfp4_moe_e2e.py]: Good catch — fixed.

File: flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    _ACC = cutlass.Float32
+    _CUTLASS_DTYPE = {
+        torch.float16: cutlass.Float16,
+        torch.bfloat16: cutlass.BFloat16,
+        torch.float32: cutlass.Float32,
+    }
+except Exception:  # pragma: no cover
+    cutlass = None
+    _ACC = None
+    _CUTLASS_DTYPE = {}
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply flashinfer/fused_moe/cute_dsl/w4a8_mxfp4_moe.py:97]: Done

File: tests/moe/test_w4a8_mxfp4_moe.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+try:
+    try:
+        from flashinfer.fused_moe import w4a8_mxfp4_moe
+    except Exception:
+        from flashinfer.fused_moe.cute_dsl import w4a8_mxfp4_moe
+    from flashinfer.utils import is_sm90a_supported
+
+    _OK = True
+except Exception:
+    _OK = False
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_mxfp4_moe.py:37]: Narrowed both `except Exception` to `except ImportError` so non-import regressions surface.

File: tests/moe/test_w4a8_mxfp4_moe.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+if __name__ == "__main__":
+    for clamp in _CLAMPS:
+        for tk in (1, 2):
+            test_w4a8_mxfp4_moe(tk, clamp)
+            print(f"PASS top_k={tk} clamp={clamp}")
+    for tk in (1, 2):
+        test_w4a8_mxfp4_moe_empty_batch(tk)
+        print(f"PASS empty_batch top_k={tk}")
+    print("ALL PASS")
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_mxfp4_moe.py]: Done.

File: tests/moe/test_w4a8_swiglu.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+)
+@pytest.mark.parametrize("M,K,I", [(128, 256, 256), (256, 512, 512), (512, 1024, 1024)])
+@pytest.mark.parametrize("clamp", _CLAMPS)
+def test_w4a8_swiglu(M, K, I, clamp):
+    dev = torch.device("cuda")
+    torch.manual_seed(0)
+    N = 2 * I  # GEMM1 N = gate|up interleaved
+    x_fp8 = (torch.randn(M, K, device=dev) / (K**0.5)).to(torch.float8_e4m3fn)
+    packed, scale, w_fp32 = _make_weight(N, K, dev, seed=1)
+    c_out = torch.empty(M, I, device=dev, dtype=torch.float8_e4m3fn)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_swiglu.py:78]: Leaving `I` as-is. E741 is explicitly in this repo's Ruff `ignore` list, so it doesn't block CI.

File: tests/moe/test_w4a8_swiglu.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+try:
+    import cutlass
+
+    from flashinfer.fused_moe import w4a8_mxfp4_grouped_gemm
+    from flashinfer.utils import is_sm90a_supported
+
+    _OK = True
+except Exception:
+    _OK = False
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply tests/moe/test_w4a8_swiglu.py:35]: Done.

**Comment by samuellees** [comment]
cc @StudyingShao Could you help review this PR, please? Appreciate~ ^ ^
